File: source/ts/src/calendar/convert.ts
import { getDailyLabels } from '../core/panchang';
import { computeSamvat, chaitraNewMoon } from '../core/samvat';
import { computeSunrise, computeSunset } from '../astronomy/sunrise';
import { getSiderealSunLongitude } from '../astronomy/sun';
import {
  resolveUtcOffset, utcDateMs, instantInCivilDay, civilDayValue, SUPPORTED_START_MS, SUPPORTED_END_MS,
} from '../utils/timezone';
import { validateLocation, validateDate } from '../utils/validation';
import { PanchangError } from '../types/errors';
import type { GeoLocation } from '../types/location';
import type {
  AyanamsaType, MasaSystem, FestivalRegion, LegacyFestivalRegion, PanchangOptions,
} from '../types/options';

/** Hindu calendar coordinates of one civil day, read at its sunrise. */
export interface HinduCalendarCoords {
  tithiName: string;
  /** 1 = Shukla Pratipada … 30 = Amavasya, not the 0..29 used internally. */
  tithi: number;
  pakshaTithi: number;
  paksha: 'shukla' | 'krishna';
  /** Carries the 'Adhika' prefix when the month is intercalary. */
  masaName: string;
  /** 0..11, in the configured `masaSystem`. */
  masaIndex: number;
  isAdhika: boolean;
  vikramSamvat: number;
  shakaSamvat: number;
  varaName: string;
  /** 0 = Sunday … 6 = Saturday. */
  varaIndex: number;
}

export interface ConvertOptions {
  /** Minutes east of UTC (`330` for IST), or an IANA zone name. */
  timezone: number | string;
  ayanamsa?: AyanamsaType;
  masaSystem?: MasaSystem;
  language?: 'en' | 'hi';
}

/** The converters read only a day's sunrise tithi, vara, lunar month and samvat, which no
 * section and no end-time search changes, so every section is off. */
function labelOptions(options: ConvertOptions): PanchangOptions {
  return { ...options, sections: [], computeEndTimes: false };
}

/**
 * Gregorian date → Hindu lunisolar coordinates at sunrise of the calendar day containing `date`
 * in `options.timezone`. West of UTC a UTC-midnight label names the day before; pass an instant
 * inside the wanted local day, such as a date `convertHinduToGregorian` returns.
 */
export function convertGregorianToHindu(
  date: Date,
  location: GeoLocation,
  options: ConvertOptions,
): HinduCalendarCoords {
  validateDate(date);
  validateLocation(location);
  const day = getDailyLabels(date, location, labelOptions(options));
  if (day === null) {
    throw new PanchangError(
      `Cannot convert ${date.toISOString()} to Hindu calendar: polar location with no sunrise`,
      'NO_SUNRISE',
    );
  }
  const tithiAtSunrise = day.tithi;
  return {
    tithiName: tithiAtSunrise.name,
    tithi: tithiAtSunrise.index + 1,
    pakshaTithi: tithiAtSunrise.number,
    paksha: tithiAtSunrise.index < 15 ? 'shukla' : 'krishna',
    masaName: day.chandramasa.name,
    masaIndex: day.chandramasa.index,
    isAdhika: day.chandramasa.isAdhika,
    vikramSamvat: day.samvat.vikramSamvat,
    shakaSamvat: day.samvat.shakaSamvat,
    varaName: day.vara.name,
    varaIndex: day.vara.index,
  };
}

/**
 * Hindu lunisolar coordinates → Gregorian dates, each the UTC midnight that falls within a local day
 * whose sunrise carries them (that date's own UTC midnight at or east of UTC, the next one west of
 * it, so read them in `options.timezone`); more than one when the tithi repeats. A `vikramSamvat`
 * that is not an integer, or whose scan window leaves 1900..2100 (or the Date range), throws
 * `PanchangError` `INVALID_DATE`.
 */
export function convertHinduToGregorian(
  coords: {
    vikramSamvat: number;
    masaIndex: number;
    paksha: 'shukla' | 'krishna';
    pakshaTithi: number;
    adhikaOnly?: boolean;
  },
  location: GeoLocation,
  options: ConvertOptions,
): Date[] {
  validateLocation(location);
  if (!Number.isInteger(coords.masaIndex) || coords.masaIndex < 0 || coords.masaIndex > 11) {
    throw new RangeError(`masaIndex must be in [0, 11], got ${coords.masaIndex}`);
  }
  if (!Number.isInteger(coords.pakshaTithi) || coords.pakshaTithi < 1 || coords.pakshaTithi > 15) {
    throw new RangeError(`pakshaTithi must be in [1, 15], got ${coords.pakshaTithi}`);
  }
  if (coords.paksha !== 'shukla' && coords.paksha !== 'krishna') {
    throw new RangeError(`paksha must be 'shukla' or 'krishna'`);
  }
  if (!Number.isInteger(coords.vikramSamvat)) {
    throw new PanchangError(`vikramSamvat must be an integer, got ${coords.vikramSamvat}`, 'INVALID_DATE');
  }

  const ceYear = coords.vikramSamvat - 57;
  const dayMs = 24 * 3600_000;
  const masaSystem = options.masaSystem ?? 'purnimanta';
  const wrapsYearEnd =
    masaSystem === 'purnimanta' && coords.masaIndex === 0 && coords.paksha === 'krishna';
  const masaMidMs = wrapsYearEnd
    ? utcDateMs(ceYear + 1, 2, 10)
    : utcDateMs(ceYear, 2, 25) + coords.masaIndex * 30 * dayMs;
  validateDate(new Date(masaMidMs - 50 * dayMs), 'any');
  validateDate(new Date(masaMidMs + 60 * dayMs), 'any');
  const targetTithi = coords.paksha === 'shukla'
    ? coords.pakshaTithi - 1
    : coords.pakshaTithi - 1 + 15;

  const dayOptions = labelOptions(options);
  const scan = (first: number, last: number): Date[] => {
    const out: Date[] = [];
    for (let day = first; day <= last; day += dayMs) {
      validateDate(new Date(day));
      const t = instantInCivilDay(day, options.timezone);
      if (t === null) continue;
      const p = getDailyLabels(new Date(t), location, dayOptions);
      if (p === null) continue;
      const tithi = p.tithi.index;
      const masa = p.chandramasa.index;
      if (tithi !== targetTithi) continue;
      if (masa !== coords.masaIndex) continue;
      if (p.samvat.vikramSamvat !== coords.vikramSamvat) continue;
      if (coords.adhikaOnly && !p.chandramasa.isAdhika) continue;
      out.push(new Date(civilDayValue(day, options.timezone)));
    }
    return out;
  };

  const out = scan(masaMidMs - 50 * dayMs, masaMidMs + 60 * dayMs);
  if (!wrapsYearEnd) return out;
  // An Adhika Chaitra's Krishna paksha opens the samvat year instead of closing it.
  const adhikaMid = utcDateMs(ceYear, 2, 25);
  return [
    ...scan(
      Math.max(adhikaMid - 50 * dayMs, SUPPORTED_START_MS),
      Math.min(adhikaMid + 60 * dayMs, SUPPORTED_END_MS),
    ),
    ...out,
  ];
}

const KALIYUGA_EPOCH_YEAR = -3101; // Kali Yuga begins 18 Feb 3102 BCE (proleptic).

/** Kali Yuga year, incrementing at Chaitra Shukla Pratipada, not on the Gregorian anniversary. */
export function getKaliYugaYear(date: Date): number {
  validateDate(date);
  const y = date.getUTCFullYear();
  const pastNewYear = date.getTime() >= chaitraNewMoon(y);
  return y - KALIYUGA_EPOCH_YEAR + (pastNewYear ? 0 : -1);
}

/**
 * The day the Hindu year begins: Chaitra Shukla Pratipada (the day containing it when no sunrise
 * does, as for Ugadi), or, in Tamil Nadu, Kerala, Punjab, Bengal, Assam and Odisha, the Mesha
 * Sankranti day by that region's rule (Odisha's Pana Sankranti: the transit's civil date, or the next
 * one when the transit falls later than 0.315 of the night after sunset). Either way the value is the
 * UTC midnight that falls within that local day, the date's own UTC midnight at or east of UTC and
 * the next one west of it, so read it in `options.timezone`.
 */
export function getHinduNewYear(
  gregorianYear: number,
  region: FestivalRegion | LegacyFestivalRegion,
  location: GeoLocation,
  options: ConvertOptions,
): Date | null {
  validateLocation(location);
  if (!Number.isInteger(gregorianYear)) {
    throw new RangeError(`gregorianYear must be integer, got ${gregorianYear}`);
  }

  const useSolarAnchor = (
    region === 'tamil' || region === 'tamil-nadu' ||
    region === 'kerala' ||
    region === 'punjab' ||
    region === 'bengal' || region === 'west-bengal' ||
    region === 'assam' ||
    region === 'odisha'
  );

  if (useSolarAnchor) {
    return findMeshaSankranti(gregorianYear, region, location, options);
  }

  return findChaitraShuklaPratipada(gregorianYear, location, options);
}

function findChaitraShuklaPratipada(
  gregorianYear: number,
  location: GeoLocation,
  options: ConvertOptions,
): Date | null {
  const amantaOptions = labelOptions({ ...options, masaSystem: 'amanta' });
  const dayMs = 24 * 3600_000;
  const last = utcDateMs(gregorianYear, 4, 15);
  let prev: { day: number; masa: number; adhika: boolean; tithi: number } | null = null;
  for (let day = utcDateMs(gregorianYear, 1, 15); day <= last; day += dayMs) {
    validateDate(new Date(day));
    const t = instantInCivilDay(day, options.timezone);
    if (t === null) continue;
    const p = getDailyLabels(new Date(t), location, amantaOptions);
    if (p === null) continue;
    const masa = p.chandramasa.index;
    const adhika = p.chandramasa.isAdhika;
    const tithi = p.tithi.index;
    if (masa === 0 && !adhika && prev !== null && (prev.masa !== 0 || prev.adhika)) {
      const kshayaPratipada =
        prev.day === day - dayMs && !prev.adhika && prev.tithi === 29 && tithi === 1;
      return new Date(civilDayValue(kshayaPratipada ? prev.day : day, options.timezone));
    }
    prev = { day, masa, adhika, tithi };
  }
  return null;
}

type MeshaDayRule = 'sankranti-day' | 'civil-day' | 'next-sunrise' | 'civil-day-plus-1' | 'night-cutoff';

/**
 * Pana Sankranti moves to the next civil date once the transit is this far into the night, in
 * thousandths of sunset to sunrise: the reference almanac's Bhubaneswar dates put the boundary
 * between 305 (2067, same date) and 327 (2028, next date).
 */
const PANA_NIGHT_CUTOFF_PER_MILLE = 315;

function meshaDayRuleFor(region: FestivalRegion | LegacyFestivalRegion): MeshaDayRule {
  switch (region) {
    case 'punjab':
      return 'civil-day';
    case 'kerala':
      return 'next-sunrise';
    case 'bengal':
    case 'west-bengal':
      return 'civil-day-plus-1';
    case 'odisha':
      return 'night-cutoff';
    default:
      return 'sankranti-day';
  }
}

function findMeshaSankranti(
  gregorianYear: number,
  region: FestivalRegion | LegacyFestivalRegion,
  location: GeoLocation,
  options: ConvertOptions,
): Date | null {
  const offsetAt = (ms: number) => resolveUtcOffset(options.timezone, new Date(ms));
  const gridOffset = offsetAt(utcDateMs(gregorianYear, 3, 1));
  const ayanamsa = options.ayanamsa ?? 'lahiri';
  const rashiAt = (ms: number) =>
    Math.floor(getSiderealSunLongitude(new Date(ms), ayanamsa) / 30) % 12;

  const dayMs = 24 * 3600_000;
  const scanStart = utcDateMs(gregorianYear, 3, 1) - gridOffset * 60_000 - dayMs;
  const scanEnd = utcDateMs(gregorianYear, 3, 20) - gridOffset * 60_000;
  let transitMs: number | null = null;
  let prevMs = scanStart;
  let prevRashi = rashiAt(prevMs);
  for (let t = scanStart + dayMs; t <= scanEnd; t += dayMs) {
    const rashi = rashiAt(t);
    if (rashi !== prevRashi && rashi === 0) {
      let lo = prevMs, hi = t;
      while (hi - lo > 1000) {
        const mid = Math.floor((lo + hi) / 2);
        if (rashiAt(mid) === prevRashi) lo = mid; else hi = mid;
      }
      transitMs = hi;
      break;
    }
    prevMs = t;
    prevRashi = rashi;
  }
  if (transitMs === null) return null;

  const civilDay = (ms: number, dayShift = 0): Date => {
    const local = new Date(ms + offsetAt(ms) * 60_000);
    return new Date(civilDayValue(utcDateMs(
      local.getUTCFullYear(), local.getUTCMonth(), local.getUTCDate() + dayShift,
    ), options.timezone));
  };

  const rule = meshaDayRuleFor(region);
  if (rule === 'civil-day') return civilDay(transitMs);
  if (rule === 'civil-day-plus-1') return civilDay(transitMs, 1);

  try {
    let dayStart = computeSunrise(new Date(transitMs - 30 * 3600_000), location);
    for (let i = 0; i < 3; i++) {
      const next = computeSunrise(computeSunset(dayStart, location), location);
      if (next.getTime() <= transitMs) dayStart = next; else break;
    }
    if (rule === 'next-sunrise') {
      const sunriseAfter = dayStart.getTime() >= transitMs
        ? dayStart
        : computeSunrise(computeSunset(dayStart, location), location);
      return civilDay(sunriseAfter.getTime());
    }
    const dayEnd = computeSunset(dayStart, location);
    if (rule === 'night-cutoff') {
      if (transitMs <= dayEnd.getTime()) return civilDay(transitMs);
      const nextSunrise = computeSunrise(dayEnd, location).getTime();
      const late = 1000 * (transitMs - dayEnd.getTime())
        > PANA_NIGHT_CUTOFF_PER_MILLE * (nextSunrise - dayEnd.getTime());
      return civilDay(late ? nextSunrise : transitMs);
    }
    const anchor = transitMs <= dayEnd.getTime()
      ? dayStart
      : computeSunrise(dayEnd, location);
    return civilDay(anchor.getTime());
  } catch (e: unknown) {
    if (!(e instanceof PanchangError && (e.code === 'NO_SUNRISE' || e.code === 'NO_SUNSET'))) {
      throw e;
    }
    return civilDay(transitMs);
  }
}

export { computeSamvat };

File: source/go/internal/calendar/convert.go
package calendar

import (
	"math"

	"github.com/ishankgupta95/panchang/source/go/v5/internal/astronomy"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/core"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/utils"
	"github.com/ishankgupta95/panchang/source/go/v5/types"
)

var AllPakshas = []Paksha{PakshaShukla, PakshaKrishna}

// convertLabelOptions is convertPanchangOptions for the converters, which read
// only each day's sunrise tithi, vara, lunar month and samvat. None of those
// depends on the optional sections or on the anga end times, so every section
// is off and the end times are not searched.
func convertLabelOptions(o ConvertOptions) types.PanchangOptions {
	opts := convertPanchangOptions(o)
	opts.Sections = core.NoSections()
	opts.SectionsGiven = true
	computeEndTimes := false
	opts.ComputeEndTimes = &computeEndTimes
	return opts
}

func ConvertGregorianToHindu(
	ctx *astronomy.EphemerisCtx,
	dateMs int64,
	location types.GeoLocation,
	options ConvertOptions,
) (HinduCalendarCoords, error) {
	if err := utils.ValidateDate(dateMs); err != nil {
		return HinduCalendarCoords{}, err
	}
	if err := utils.ValidateLocation(location); err != nil {
		return HinduCalendarCoords{}, err
	}
	day, ok, err := core.GetDailyLabels(ctx, dateMs, location, convertLabelOptions(options))
	if err != nil {
		return HinduCalendarCoords{}, err
	}
	if !ok {
		return HinduCalendarCoords{}, types.NewPanchangError(
			"Cannot convert "+types.Date(dateMs).ISOString()+
				" to Hindu calendar: polar location with no sunrise",
			types.ErrNoSunrise)
	}
	tithiAtSunrise := day.Tithi
	paksha := PakshaKrishna
	if tithiAtSunrise.Index < 15 {
		paksha = PakshaShukla
	}
	return HinduCalendarCoords{
		TithiName:    tithiAtSunrise.Name,
		Tithi:        tithiAtSunrise.Index + 1,
		PakshaTithi:  tithiAtSunrise.Number,
		Paksha:       paksha,
		MasaName:     day.Chandramasa.Name,
		MasaIndex:    day.Chandramasa.Index,
		IsAdhika:     day.Chandramasa.IsAdhika,
		VikramSamvat: day.Samvat.VikramSamvat,
		ShakaSamvat:  day.Samvat.ShakaSamvat,
		VaraName:     day.Vara.Name,
		VaraIndex:    day.Vara.Index,
	}, nil
}

func ConvertHinduToGregorian(
	ctx *astronomy.EphemerisCtx,
	coords HinduDateCoords,
	location types.GeoLocation,
	options ConvertOptions,
) ([]types.JSDate, error) {
	if err := utils.ValidateLocation(location); err != nil {
		return nil, err
	}
	if coords.MasaIndex < 0 || coords.MasaIndex > 11 {
		return nil, types.Codef(types.ErrInvalidInput,
			"masaIndex must be in [0, 11], got %d", coords.MasaIndex)
	}
	if coords.PakshaTithi < 1 || coords.PakshaTithi > 15 {
		return nil, types.Codef(types.ErrInvalidInput,
			"pakshaTithi must be in [1, 15], got %d", coords.PakshaTithi)
	}
	if coords.Paksha != PakshaShukla && coords.Paksha != PakshaKrishna {
		return nil, types.Codef(types.ErrInvalidInput,
			"paksha must be 'shukla' or 'krishna'")
	}

	ceYear := coords.VikramSamvat - 57
	masaSystem := convertResolvedMasaSystem(options)
	wrapsYearEnd := masaSystem == types.Purnimanta &&
		coords.MasaIndex == 0 && coords.Paksha == PakshaKrishna
	var masaMidMs int64
	if wrapsYearEnd {
		masaMidMs = utils.UtcDateMs(ceYear+1, 2, 10)
	} else {
		masaMidMs = utils.UtcDateMs(ceYear, 2, 25) + int64(coords.MasaIndex)*30*dayMs
	}
	targetTithi := coords.PakshaTithi - 1
	if coords.Paksha == PakshaKrishna {
		targetTithi += 15
	}

	opts := convertLabelOptions(options)
	scan := func(firstMs, lastMs int64) ([]types.JSDate, error) {
		out := []types.JSDate{}
		for day := firstMs; day <= lastMs; day += dayMs {
			if err := utils.ValidateDate(day); err != nil {
				return nil, err
			}
			t, onDay, err := utils.InstantInCivilDay(day, options.Timezone)
			if err != nil {
				return nil, err
			}
			if !onDay {
				continue
			}
			p, ok, err := core.GetDailyLabels(ctx, t, location, opts)
			if err != nil {
				return nil, err
			}
			if !ok {
				continue
			}
			tithi := p.Tithi.Index
			masa := p.Chandramasa.Index
			if tithi != targetTithi {
				continue
			}
			if masa != coords.MasaIndex {
				continue
			}
			if p.Samvat.VikramSamvat != coords.VikramSamvat {
				continue
			}
			if coords.AdhikaOnly && !p.Chandramasa.IsAdhika {
				continue
			}
			value, err := utils.CivilDayValue(day, options.Timezone)
			if err != nil {
				return nil, err
			}
			out = append(out, types.Date(value))
		}
		return out, nil
	}

	out, err := scan(masaMidMs-50*dayMs, masaMidMs+60*dayMs)
	if err != nil || !wrapsYearEnd {
		return out, err
	}
	// An Adhika Chaitra's Krishna paksha opens the samvat year instead of closing it.
	adhikaMidMs := utils.UtcDateMs(ceYear, 2, 25)
	adhikaFirstMs, adhikaLastMs := adhikaMidMs-50*dayMs, adhikaMidMs+60*dayMs
	if adhikaFirstMs < utils.SupportedStartMs {
		adhikaFirstMs = utils.SupportedStartMs
	}
	if adhikaLastMs > utils.SupportedEndMs {
		adhikaLastMs = utils.SupportedEndMs
	}
	adhika, err := scan(adhikaFirstMs, adhikaLastMs)
	if err != nil {
		return nil, err
	}
	return append(adhika, out...), nil
}

const kaliyugaEpochYear = -3101

func GetKaliYugaYear(ctx *astronomy.EphemerisCtx, dateMs int64) (int, error) {
	if err := utils.ValidateDate(dateMs); err != nil {
		return 0, err
	}
	y := types.Date(dateMs).UTCFullYear()
	newMoonMs, err := core.ChaitraNewMoon(ctx, y)
	if err != nil {
		return 0, err
	}
	if dateMs >= newMoonMs {
		return y - kaliyugaEpochYear, nil
	}
	return y - kaliyugaEpochYear - 1, nil
}

func GetHinduNewYear(
	ctx *astronomy.EphemerisCtx,
	gregorianYear int,
	region types.FestivalRegion,
	location types.GeoLocation,
	options ConvertOptions,
) (types.JSDate, bool, error) {
	if err := utils.ValidateLocation(location); err != nil {
		return 0, false, err
	}

	useSolarAnchor := region == types.LegacyRegionTamil || region == types.RegionTamilNadu ||
		region == types.RegionKerala ||
		region == types.RegionPunjab ||
		region == types.LegacyRegionBengal || region == types.RegionWestBengal ||
		region == types.RegionAssam ||
		region == types.RegionOdisha

	if useSolarAnchor {
		return findMeshaSankranti(ctx, gregorianYear, region, location, options)
	}

	return findChaitraShuklaPratipada(ctx, gregorianYear, location, options)
}

type chandraMasaState struct {
	masa   int
	adhika bool
	have   bool
	dayMs  int64
	tithi  int
}

func findChaitraShuklaPratipada(
	ctx *astronomy.EphemerisCtx,
	gregorianYear int,
	location types.GeoLocation,
	options ConvertOptions,
) (types.JSDate, bool, error) {
	amantaOptions := convertLabelOptions(options)
	amantaOptions.MasaSystem = types.Amanta
	lastMs := utils.UtcDateMs(gregorianYear, 4, 15)
	var prev chandraMasaState
	for day := utils.UtcDateMs(gregorianYear, 1, 15); day <= lastMs; day += dayMs {
		if err := utils.ValidateDate(day); err != nil {
			return 0, false, err
		}
		t, onDay, err := utils.InstantInCivilDay(day, options.Timezone)
		if err != nil {
			return 0, false, err
		}
		if !onDay {
			continue
		}
		p, ok, err := core.GetDailyLabels(ctx, t, location, amantaOptions)
		if err != nil {
			return 0, false, err
		}
		if !ok {
			continue
		}
		masa := p.Chandramasa.Index
		adhika := p.Chandramasa.IsAdhika
		tithi := p.Tithi.Index
		if masa == 0 && !adhika && prev.have && (prev.masa != 0 || prev.adhika) {
			kshayaPratipada := prev.dayMs == day-dayMs && !prev.adhika && prev.tithi == 29 && tithi == 1
			newYear := day
			if kshayaPratipada {
				newYear = prev.dayMs
			}
			value, err := utils.CivilDayValue(newYear, options.Timezone)
			if err != nil {
				return 0, false, err
			}
			return types.Date(value), true, nil
		}
		prev = chandraMasaState{masa: masa, adhika: adhika, have: true, dayMs: day, tithi: tithi}
	}
	return 0, false, nil
}

type meshaDayRule string

const (
	meshaSankrantiDay  meshaDayRule = "sankranti-day"
	meshaCivilDay      meshaDayRule = "civil-day"
	meshaNextSunrise   meshaDayRule = "next-sunrise"
	meshaCivilDayPlus1 meshaDayRule = "civil-day-plus-1"
	meshaNightCutoff   meshaDayRule = "night-cutoff"
)

// panaNightCutoffPerMille is how far into the night, in thousandths of sunset
// to sunrise, a transit moves Pana Sankranti to the next civil date: the
// reference almanac's Bhubaneswar dates put it between 305 (2067, same date)
// and 327 (2028, next date).
const panaNightCutoffPerMille int64 = 315

func meshaDayRuleFor(region types.FestivalRegion) meshaDayRule {
	switch region {
	case types.RegionPunjab:
		return meshaCivilDay
	case types.RegionKerala:
		return meshaNextSunrise
	case types.LegacyRegionBengal, types.RegionWestBengal:
		return meshaCivilDayPlus1
	case types.RegionOdisha:
		return meshaNightCutoff
	default:
		return meshaSankrantiDay
	}
}

func findMeshaSankranti(
	ctx *astronomy.EphemerisCtx,
	gregorianYear int,
	region types.FestivalRegion,
	location types.GeoLocation,
	options ConvertOptions,
) (types.JSDate, bool, error) {
	gridOffset, err := utils.ResolveUtcOffset(options.Timezone, utils.UtcDateMs(gregorianYear, 3, 1))
	if err != nil {
		return 0, false, err
	}
	ayanamsa := convertResolvedAyanamsa(options)
	rashiAt := func(ms int64) (int, error) {
		lon, err := astronomy.GetSiderealSunLongitude(ctx, ms, ayanamsa)
		if err != nil {
			return 0, err
		}
		return int(math.Floor(lon/30)) % 12, nil
	}

	scanStart := utils.UtcDateMs(gregorianYear, 3, 1) - int64(gridOffset)*60_000 - dayMs
	scanEnd := utils.UtcDateMs(gregorianYear, 3, 20) - int64(gridOffset)*60_000
	transitMs, haveTransit := int64(0), false
	prevMs := scanStart
	prevRashi, err := rashiAt(prevMs)
	if err != nil {
		return 0, false, err
	}
	for t := scanStart + dayMs; t <= scanEnd; t += dayMs {
		rashi, err := rashiAt(t)
		if err != nil {
			return 0, false, err
		}
		if rashi != prevRashi && rashi == 0 {
			lo, hi := prevMs, t
			for hi-lo > 1000 {
				mid := (lo + hi) >> 1
				r, err := rashiAt(mid)
				if err != nil {
					return 0, false, err
				}
				if r == prevRashi {
					lo = mid
				} else {
					hi = mid
				}
			}
			transitMs, haveTransit = hi, true
			break
		}
		prevMs, prevRashi = t, rashi
	}
	if !haveTransit {
		return 0, false, nil
	}

	civilDay := func(ms int64, dayShift int) (types.JSDate, bool, error) {
		offset, err := utils.ResolveUtcOffset(options.Timezone, ms)
		if err != nil {
			return 0, false, err
		}
		local := types.Date(utils.UtcToLocalDisplay(ms, offset))
		value, err := utils.CivilDayValue(
			utils.UtcDateMs(local.UTCFullYear(), local.UTCMonth(), local.UTCDate()+dayShift), options.Timezone)
		if err != nil {
			return 0, false, err
		}
		return types.Date(value), true, nil
	}

	rule := meshaDayRuleFor(region)
	if rule == meshaCivilDay {
		return civilDay(transitMs, 0)
	}
	if rule == meshaCivilDayPlus1 {
		return civilDay(transitMs, 1)
	}

	day, err := meshaSunriseAnchor(ctx, transitMs, rule, location)
	if err != nil {
		if !isPolarRiseSetError(err) {
			return 0, false, err
		}
		return civilDay(transitMs, 0)
	}
	return civilDay(day, 0)
}

func meshaSunriseAnchor(
	ctx *astronomy.EphemerisCtx,
	transitMs int64,
	rule meshaDayRule,
	location types.GeoLocation,
) (int64, error) {
	dayStart, err := astronomy.ComputeSunrise(ctx, transitMs-30*3600_000, location,
		astronomy.DefaultRiseSetLimitDays)
	if err != nil {
		return 0, err
	}
	for i := 0; i < 3; i++ {
		sunset, err := astronomy.ComputeSunset(ctx, dayStart, location,
			astronomy.DefaultRiseSetLimitDays)
		if err != nil {
			return 0, err
		}
		next, err := astronomy.ComputeSunrise(ctx, sunset, location,
			astronomy.DefaultRiseSetLimitDays)
		if err != nil {
			return 0, err
		}
		if next <= transitMs {
			dayStart = next
		} else {
			break
		}
	}
	if rule == meshaNextSunrise {
		if dayStart >= transitMs {
			return dayStart, nil
		}
		sunset, err := astronomy.ComputeSunset(ctx, dayStart, location,
			astronomy.DefaultRiseSetLimitDays)
		if err != nil {
			return 0, err
		}
		return astronomy.ComputeSunrise(ctx, sunset, location, astronomy.DefaultRiseSetLimitDays)
	}
	dayEnd, err := astronomy.ComputeSunset(ctx, dayStart, location,
		astronomy.DefaultRiseSetLimitDays)
	if err != nil {
		return 0, err
	}
	if rule == meshaNightCutoff {
		if transitMs <= dayEnd {
			return transitMs, nil
		}
		nextSunrise, err := astronomy.ComputeSunrise(ctx, dayEnd, location, astronomy.DefaultRiseSetLimitDays)
		if err != nil {
			return 0, err
		}
		if 1000*(transitMs-dayEnd) > panaNightCutoffPerMille*(nextSunrise-dayEnd) {
			return nextSunrise, nil
		}
		return transitMs, nil
	}
	if transitMs <= dayEnd {
		return dayStart, nil
	}
	return astronomy.ComputeSunrise(ctx, dayEnd, location, astronomy.DefaultRiseSetLimitDays)
}

func ComputeSamvat(ctx *astronomy.EphemerisCtx, dateMs int64) (types.SamvatInfo, error) {
	return core.ComputeSamvat(ctx, dateMs)
}
